### DISCUSSION TITLE
User Accounts/Improved chat user identity

### DISCUSSION THREAD

**Discussion by gabek** [owncast/owncast#440] [Ideas]
I wanted to open a conversation around the future of the chat functionality, and throw around some ideas on how to tighten up user identity.

### Right now
**Pro:** You don't need to do anything to start chatting.
**Con:** Anything goes.  You can set your name to whatever you want, including the same name as somebody else.  There's no way to differentiate one person from another.  There's no metadata around a user and nothing uniquely identifying them.  Because we have no solid identity it will make chat moderation more difficult.

### What I'd like
Fix the downsides while keeping the simplicity that we have.

### Idea
I've been thinking of an approach around the idea of some kind of "virtual account".  In a nutshell this is the idea:

1. A user visits the Owncast web interface for the first time.
1. The web frontend makes an API call telling the backend to create a new chat user for this new person.
1. Some basic metadata around this user is saved in the database: a user id, ip address, a random color, the date the user was created, and their auto-generated username.
1. Why IP address?  Because we'll need to eventually build chat moderation and we need something more than just the id to use for blacklisting users.
1. A cookie is set with this identifier.
1. The user is free to change their name anytime they like to anything they like.  However, their identity will stay the same.  If you hover over the user's message you'll see the date the user was created, and their color will stay the same regardless of the username.  So for example: If somebody joins the chat and changes their name to "gabe", it could say "Created Nov 15. Name last changed 1 minute ago" and show in orange.  Where if you hover over the _real_ "gabe" it would say "Created Mar 3. Name last changed three months ago" and show in blue.
1. In this scenario there can be any number of "gabe" in the chat, but it's easy to determine the differences.
1. The next time the user visits the web interface the cookie will have the user id and get sent along with chat messages, using the same "virtual account" that has the metadata in the backend.
1. If a user were to change computers, browsers or just loose their cookie, it's no big deal.  A new one would be created and they'd start over.

### Bonus idea

Surface the ID of the user as something portable so they can move it between machines or browsers.  Have some kind of way to put that ID in and restore your identity.

There'd be a lot of changes required to make something like this happen, but I don't think it'd be overly complex.

## Thoughts?

I'd love to hear any other suggestions along the lines of keeping a non-authenticated chat experience, but improving identity.

**Comment by gabek** [comment]
See https://github.com/owncast/owncast/issues/464 for suggestions around having full user accounts.

**Comment by jeyemwey** [comment]
I’m here to throw JSON Web Token into the mix. With JWT, you can sign the user ids and revoke that signature. Maybe, this will help us later?

**Comment by ForestJohnson** [comment]
one thing that bugged me personally, this is just my 2c, it seems wierd that when a user joins, they are assigned a username in the UI, like "User 22" or whatever but when you call `/api/admin/clients` their username will be `null`. Seems inconsistent, no ? As a naive consumer of the owncast api, even if I wanted to determine which username the client had assigned them, I would have no way of doing that.

**Comment by ForestJohnson** [comment]
I will say it might be a good idea to reserve names for the oldest registered user so if you get a new `gabe` user they would show up as `gabe[2]` or something

**Comment by gabek** [comment]
@ginger did some awesome mocks, so posting them here so they're with the rest of the discussion.

![image](https://user-images.githubusercontent.com/414923/112561816-2d05dd00-8d93-11eb-9a74-3fc52765c1ed.png)

**Comment by gabek** [comment]
It's time to make the next moves with Owncast chat to allow for moderation, functionality, and flexibility.  Read the above thread if you're just now getting to this post so you can see the details and past comments.

## Things we want

1. Chat user uniqueness (you are you, and can resume being the same you).
2. Chat user identification (you have user info and metadata).
3. Improved moderation to be able to ban users via IP or unique identifier.

Here's how we can iteratively work through this.  Each step has less detail than the previous, but we will learn each step of the way and decide on future choices.

### Step 1.  Shadow accounts.

1. Build the "shadow account" functionality (read above in this discussion) where the chat user gets an implicit account created on the backend for unique identification, with no auth required or available.
2. This allows for functionality that requires uniqueness.  Examples: Only allow a person to vote for a poll once, be able to ban a user from chat.
3. The downside is if a person were to switch browsers or join from another device we wouldn't be able to link this "shadow account".
4. Even with just this in place we could add the ability for people to moderate chat.  As long as the server admin can identify users uniquely, and we have a way to know it's them, we can give them privileges.

### Step 2. Small scale authentication, allowing the admin to auth.

1. Allow the admin to "upgrade" or "verify" their "shadow account" by authenticating against their server. This can be done when on their own server or another Owncast server by allowing Owncast to operate as an IndieAuth provider https://indieweb.org/IndieAuth. 
2. With the admittedly small target group of owncast server admins, using Owncast server instances as an authentication provider, we can start small and allow this group to be authenticated working through the rough edges and deciding what is to come for step 3.
3. Example: Gabe, who runs goth.land visits live.w0rmh0le.net.  He has an auto-generated "shadow account" on live.w0rmh0le.net and knows it's Gabe, but it's not verified.  He starts the the auth flow, types in "goth.land" and live.w0rmh0le.net verifies Gabe is verified by goth.land, and that "shadow account" is now an authorized, "verified" account.
4. It also fixes a short term issue where jerk chat users use the same chat username as the admin of the server, so this would allow the admin to be "verified".
5. As a side effect it would promote the Owncast community by streamers being identified in other streamers' chats.

### Step 3: Additional providers and full auth for everyone.

Once we iron out any issues and polish the auth flow we can add other authentication providers for the general chat audience and allow everyone to verify their identity by upgrading their "shadow accounts" and everyone will be able to be identified and auth'ed.

After this we can do fancy things like:
* Allow people's auth to grant them access to the stream or chat.
* Display metadata about chat users.
* Support moderators.
* Have auth'ed-only chat.
* And more!

**Reply by geekgonecrazy** [reply]
Great!  Just to confirm how you are thinking.. Would shadow account be in form of a session and cookie I guess and tied to a browser? I assume so since switching browsers was mentioned. 

I’m curious.. with indie auth would we actually even be able to tell it’s another owncast server? This is a really interesting idea though

**Reply by gabek** [reply]
Yup, it would be linked to a browser for that first version of it.

And technically it would work with anywhere that supports IndieAuth, but for the sake of "verifying" an admin account, it makes sense to talk about it a bit narrower.  But it's not limited to that.  Hopefully there's a lightweight way we can implement it, allowing the admin to still just use the stream key to authenticate, but I haven't looked into specifics.  @aaronpk suggested this idea and I thought it was a good fit for an iterative approach in how we allow people to authenticate.

**Reply by aaronpk** [reply]
Looks like a great plan! 

I'd recommend actually as part of step 1, automatically creating a "slightly upgraded shadow account" for the server admin so that when they are logged in they will always have the same chat name automatically. It shouldn't require any additional actions for the admin user to take. When they are logged in to their own server, they would just already have their chat user account set.

Step 2 would include the part where owncast users could log in to each others' owncast sites (using IndieAuth, and of course not limited to OwnCast sites since there's no real reason to limit it, but that would obviously be the most common case here.)

**Comment by eayin2** [comment]
For my use case I need just a user based authentication. Only registered and activated users should be allowed to see the stream. Looking at this project I think it uses preact for the javascript web framework. For that use case, wouldn't it be enough to simply add _Okta_ to do this project? Not sure how much it covers your use case, but I think this should be easily integrated within a few days.

**Reply by aaronpk** [reply]
In that case I would highly recommend setting up authentication at your web server layer rather than in this project itself. I actually wrote up a tutorial on doing exactly this with Okta, and you can use this to protect any application, including OwnCast this way:

https://developer.okta.com/blog/2018/08/28/nginx-auth-request

**Comment by drizzt09** [comment]
Any plans still for step 3. Would like my general users to be able to register their usernames in chat. 
Those mock up screenshowing Google and Twitter options would be nice. Does these ever work?

**Reply by gabek** [reply]
Yes. Please check the documentation before posting, you'll often find your answers. See the documentation for chat authentication. https://owncast.online/docs/chat/chat-authentication/

**Comment by drizzt09** [comment]
I have read that. It tells me nothing. It says I can use indieauth which is
built in and that works for me as admin.  This does not work for any of my
guest viewers. None of them have an owncast server or an indieauth server
of their own. And that all looks like step2. Not 3.

Your doc also says I can use fedeverse. Which basically seems like the same
thing. None of my viewers have fedeverse.

I have a user who got random name. Manually changed his name. I see him in
admin panel and set to mod. He doesn't have mod abilities.
